From e91271d3d4ba1224c80409fd2b8250b8e704a421 Mon Sep 17 00:00:00 2001
From: Amir Vadai <avadai@redhat.com>
Date: Thu, 11 Sep 2014 11:59:40 -0400
Subject: [infiniband] mlx4: Fix error return code

Message-id: <7db055890f849e673d8ff9d43cf55730595c348d.1410425016.git.avadai@redhat.com>
Patchwork-id: 93675
O-Subject: [Patch RHEL7.1 V1 044/197] IB/mlx4: Fix error return code
Bugzilla: 1107617 1107618 1107619 1107620
RH-Acked-by: Jes Sorensen <Jes.Sorensen@redhat.com>

From: Julia Lawall <Julia.Lawall@lip6.fr>

Set the return variable to an error code as done elsewhere in the function.

A simplified version of the semantic match that finds this problem is as
follows: (http://coccinelle.lip6.fr/)

// <smpl>
(
if@p1 (\(ret < 0\|ret != 0\))
 { ... return ret; }
|
ret@p1 = 0
)
... when != ret = e1
    when != &ret
*if(...)
{
  ... when != ret = e2
      when forall
 return ret;
}

// </smpl>

Signed-off-by: Julia Lawall <Julia.Lawall@lip6.fr>
Signed-off-by: Roland Dreier <roland@purestorage.com>
(cherry picked from commit af2e2e35a23e4aeecfe4332a7140c81e0f09b7e3)
Signed-off-by: Amir Vadai <avadai@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/infiniband/hw/mlx4/sysfs.c b/drivers/infiniband/hw/mlx4/sysfs.c
index 97516eb..db2ea31 100644
--- a/drivers/infiniband/hw/mlx4/sysfs.c
+++ b/drivers/infiniband/hw/mlx4/sysfs.c
@@ -582,8 +582,10 @@ static int add_port(struct mlx4_ib_dev *dev, int port_num, int slave)
  p->pkey_group.attrs =
   alloc_group_attrs(show_port_pkey, store_port_pkey,
       dev->dev->caps.pkey_table_len[port_num]);
- if (!p->pkey_group.attrs)
+ if (!p->pkey_group.attrs) {
+  ret = -ENOMEM;
   goto err_alloc;
+ }
 
  ret = sysfs_create_group(&p->kobj, &p->pkey_group);
  if (ret)
@@ -591,8 +593,10 @@ static int add_port(struct mlx4_ib_dev *dev, int port_num, int slave)
 
  p->gid_group.name  = "gid_idx";
  p->gid_group.attrs = alloc_group_attrs(show_port_gid_idx, NULL, 1);
- if (!p->gid_group.attrs)
+ if (!p->gid_group.attrs) {
+  ret = -ENOMEM;
   goto err_free_pkey;
+ }
 
  ret = sysfs_create_group(&p->kobj, &p->gid_group);
  if (ret)
-- 
1.7.1